src/sentry/debug_files/artifact_bundles.py

```diff
@@ -0,0 +1,114 @@
+    - If a key already exists with a higher `date_last_modified`, it is ignored
+      as that indicates out-of-order calls of this method. We always want the
+      most up-to-date bundle to win.
+    - Otherwise, the key is created or updated to point to the `artifact_bundle`.
+    """
+
+    # First: Mark the bundles as `is_indexing` using a compare-and-swap like update
+    started_artifact_bundles = []
+    for artifact_bundle in artifact_bundles:
+        did_start_indexing = ArtifactBundle.objects.filter(
+            id=artifact_bundle.id, indexing_state="needs_indexing"
+        ).update(indexing_state="is_indexing")
+        if did_start_indexing:
+            started_artifact_bundles.append(artifact_bundle)
+        else:
+            # a different job is already taking care of this bundle
+            metrics.incr("artifact_bundle_indexing.conflict")
+    artifact_bundles = started_artifact_bundles
+
+    if not artifact_bundles:
+        return
+
+    # Then: Merge the bundles into an in-memory index
+    files_to_index: Mapping[str, ArtifactBundle] = {}
+
+    for artifact_bundle in artifact_bundles:
+        archive = ArtifactBundleArchive(artifact_bundle.file.getfile(), build_memory_map=False)
+        try:
+            for info in archive.manifest.get("files", {}).values():
+                url = info.get("url")
+                if not url:
+                    continue
+
+                indexed = files_to_index.get(url)
+                if not indexed or indexed.date_last_modified < artifact_bundle.date_last_modified:
+                    files_to_index[url] = artifact_bundle
+
+        finally:
+            archive.close()
+
+    # this is used for partitioning
+    date_touched = datetime.now()
+
+    # And finally, commit the intended changes to the database, carefully merging
+    # with the state that already exists in the database.
+    indexed_urls = 0
+    for url, artifact_bundle in files_to_index:
+        key = {
+            "organization_id": artifact_bundle.organization_id,
+            "release": release,
+            "dist": dist,
+            "url": url,
+        }
+        value = {
+            "date_last_modified": artifact_bundle.date_last_modified,
+            "date_touched": date_touched,
+            "artifact_bundle_id": artifact_bundle.id,
+        }
+
+        # TODO: we are doing at least one update for each file
+        # is it even possible to do these operations in a batched way?
+        # what we want is:
+        # - create is `url` does not exist
+        # - update if `date_last_modified` is less
+        # - otherwise ignore
+
+        did_update = ArtifactUrlIndex.objects.filter(
+            **key, date_last_modified__lt=artifact_bundle.date_last_modified
+        ).update(**value)
+        if not did_update:
+            # TODO: on duplicate key continue?
+            ArtifactUrlIndex.objects.create(
+                **key,
+                **value,
+            )
```

Comment:
You could use a transaction in the following way:
```suggestion
            with atomic_transaction(
                using=(
                    router.db_for_write(ArtifactBundleIndex),
                )
            ):
	        did_update = ArtifactBundleIndex.objects.filter(
	            **key, Q(date_last_modified__lt=artifact_bundle.date_last_modified) | Q(date_last_modified=artifact_bundle.date_last_modified, artifact_bundle_id__lt=artifact_bundle_id)
	        ).update(**value)
	        if not did_update:
	            ArtifactBundleIndex.objects.create(
	                **key,
	                **value,
	            )
```

Also here, I would put the comparison to guarantee total order.

Reply:
I personally would not use dict destructuring since it's not very nice to read but it's more a matter of preference.